rfcs/45-stageless.md

```diff
@@ -0,0 +1,564 @@
+
+1. **Strict ordering constraints:** `strictly_before` and `strictly_after`
+   1. A system cannot be scheduled until "strictly before" systems must have been completed this frame.
+   2. Simple and explicit.
+   3. Can cause unneccessary blocking, particularly when systems are configured at a high-level.
+2. **If-needed ordering constraints:** `.before` and `.after`
+   1. A system cannot be scheduled until any "before" systems that it is incompatible with have completed this frame.
+   2. In 99% of cases, this is the desired behavior. Unless you are using interior mutability, systems that are compatible will always be **commutative**: their ordering doesn't matter.
+
+Applying an ordering constraint to or from a label causes a ordering constraint to be created between all individual members of that label.
+If an ordering is defined relative to a non-existent system or label, it will have no effect, emitting a warning.
+This relatively gentle failure mode is important to ensure that plugins can order their systems with relatively strong assumptions that the default system labels exist, but continue to (mostly) work if those systems or labels are not present.
+
+In addition to the `.before` and `.after` methods, you can use **system chains** to create very simple linear dependencies between the succesive members of an array of systems.
+
+When discussing system ordering, it is particularly important to call out the `flush_commands` system.
+This **exclusive system** (meaning, it can modify the entire `World` in arbitrary ways and cannot be run in parallel with other systems) collects all created commands and applies them to the `World`.
+
+Commands (commonly used to spawn and despawn entities or add and remove components) will not take effect until this system is run, so be sure that you run a copy of the `flush_commands` system before relying on the result of a command!
+
+This pattern is so common that a special form of ordering constraint exists for it: **command-flushed ordering constraints**.
+If system `A` is `before_and_flush` system `B`, the schedule will be unsatisfiable unless there is an intervening `flush_commands` system.
+Note that **this does not insert new copies of a `flush_commands` system**: instead, it is merely used to verify that your schedule has been set up correctly according the specified constraint.
+
+### Run criteria
+
+While ordering constraints determine *when* a system will run, **run criteria** will determine *if* it will run at all.
+A run criteria is a special kind of system, which can read data from the `World` and returns a boolean value.
+If its output is `true`, the system it is attached to will run during this pass of the `Schedule`;
+if it is `false`, the system will be skipped.
+Systems that are skipped are considered completed for the purposes of ordering constraints.
+If for some reason run criteria are themselves skipped, they are considered to have returned `true`.

```

Comment:
The reason for this choice is that "return true" is effectively the safe, default behavior. I'd rather have more graceful failures in the case of this weird edge case, especially since a single run criteria returning false will skip the whole system.

Reply:
Yes, it seems safer (not as in soundness/ub, but as in "less likely to lead to unexpected logic bugs") to assume users want their logic to run, than to assume they want it to be skipped. It's a niche scenario anyway.